File: electron/sqlite/queries/members.js
/**
 * 成员相关 SQLite 查询模块
 *
 * 实现 members、worker_teams、worker_transfer_records、project_members 四张表的 CRUD 操作。
 */
import log from 'electron-log';
import { tryGetSqlite, rowToCamel, toSqliteValue } from './helpers';
// ═══════════════════════════════════════════════════════════════════════════════
// Members — 列映射
// ═══════════════════════════════════════════════════════════════════════════════
var MEM_COLUMNS = {
    id: 'id',
    name: 'name',
    phone: 'phone',
    email: 'email',
    memberType: 'member_type',
    role: 'role',
    workerType: 'worker_type',
    idCard: 'id_card',
    idCardFront: 'id_card_front',
    idCardBack: 'id_card_back',
    gender: 'gender',
    ethnicity: 'ethnicity',
    birthDate: 'birth_date',
    idCardAddress: 'id_card_address',
    contractFile: 'contract_file',
    contractFileType: 'contract_file_type',
    baseSalary: 'base_salary',
    socialSecurityPersonal: 'social_security_personal',
    socialSecurityCompany: 'social_security_company',
    housingFund: 'housing_fund',
    housingFundPersonal: 'housing_fund_personal',
    otherAllowances: 'other_allowances',
    companyCoversSocial: 'company_covers_social',
    teamId: 'team_id',
    dailyWage: 'daily_wage',
    entryDate: 'entry_date',
    expectedLeaveDate: 'expected_leave_date',
    actualLeaveDate: 'actual_leave_date',
    wageBankAccount: 'wage_bank_account',
    wageBankName: 'wage_bank_name',
    threeLevelEducation: 'three_level_education',
    safetyTrainingFile: 'safety_training_file',
    healthReportFile: 'health_report_file',
    specialCertificateFile: 'special_certificate_file',
    status: 'status',
    leaveDate: 'leave_date',
    reentryDate: 'reentry_date',
    remarks: 'remarks',
    departmentId: 'department_id',
    position: 'position',
    projectId: 'project_id',
    isTeamLeader: 'is_team_leader',
    createdAt: 'created_at',
};
var MEM_INSERT_COLS = Object.values(MEM_COLUMNS).filter(function (c) { return c !== 'id'; });
var MEM_INSERT_SQL = "INSERT INTO members (".concat(MEM_INSERT_COLS.map(function (c) { return "\"".concat(c, "\""); }).join(', '), ") VALUES (").concat(MEM_INSERT_COLS.map(function () { return '?'; }).join(', '), ")");
// ═══════════════════════════════════════════════════════════════════════════════
// Worker Teams — 列映射
// ═══════════════════════════════════════════════════════════════════════════════
var TEAM_COLUMNS = {
    id: 'id',
    name: 'name',
    projectId: 'project_id',
    leaderId: 'leader_id',
    createdAt: 'created_at',
    updatedAt: 'updated_at',
};
var TEAM_INSERT_COLS = Object.values(TEAM_COLUMNS).filter(function (c) { return c !== 'id'; });
var TEAM_INSERT_SQL = "INSERT INTO worker_teams (".concat(TEAM_INSERT_COLS.map(function (c) { return "\"".concat(c, "\""); }).join(', '), ") VALUES (").concat(TEAM_INSERT_COLS.map(function () { return '?'; }).join(', '), ")");
// ═══════════════════════════════════════════════════════════════════════════════
// Worker Transfer Records — 列映射
// ═══════════════════════════════════════════════════════════════════════════════
var TRANSFER_COLUMNS = {
    id: 'id',
    workerId: 'worker_id',
    fromTeamId: 'from_team_id',
    toTeamId: 'to_team_id',
    fromProjectId: 'from_project_id',
    toProjectId: 'to_project_id',
    transferDate: 'transfer_date',
    reason: 'reason',
    createdAt: 'created_at',
};
var TRANSFER_INSERT_COLS = Object.values(TRANSFER_COLUMNS).filter(function (c) { return c !== 'id'; });
var TRANSFER_INSERT_SQL = "INSERT INTO worker_transfer_records (".concat(TRANSFER_INSERT_COLS.map(function (c) { return "\"".concat(c, "\""); }).join(', '), ") VALUES (").concat(TRANSFER_INSERT_COLS.map(function () { return '?'; }).join(', '), ")");
// ═══════════════════════════════════════════════════════════════════════════════
// Project Members — 列映射
// ═══════════════════════════════════════════════════════════════════════════════
var PM_COLUMNS = {
    id: 'id',
    projectId: 'project_id',
    memberId: 'member_id',
    joinedAt: 'joined_at',
};
var PM_INSERT_COLS = Object.values(PM_COLUMNS).filter(function (c) { return c !== 'id'; });
var PM_INSERT_SQL = "INSERT INTO project_members (".concat(PM_INSERT_COLS.map(function (c) { return "\"".concat(c, "\""); }).join(', '), ") VALUES (").concat(PM_INSERT_COLS.map(function () { return '?'; }).join(', '), ")");
// ═══════════════════════════════════════════════════════════════════════════════
// 辅助：通用 INSERT 参数生成
// ═══════════════════════════════════════════════════════════════════════════════
function toInsertParams(columns, insertCols, obj) {
    return insertCols.map(function (col) {
        var _a;
        var jsonKey = (_a = Object.entries(columns).find(function (_a) {
            var c = _a[1];
            return c === col;
        })) === null || _a === void 0 ? void 0 : _a[0];
        if (!jsonKey)
            return null;
        return toSqliteValue(obj[jsonKey]);
    });
}
function toUpdateSet(columns, changes, excludeKeys) {
    if (excludeKeys === void 0) { excludeKeys = []; }
    var setClauses = [];
    var params = [];
    for (var _i = 0, _a = Object.entries(changes); _i < _a.length; _i++) {
        var _b = _a[_i], jsonKey = _b[0], value = _b[1];
        if (excludeKeys.includes(jsonKey))
            continue;
        var col = columns[jsonKey];
        if (!col)
            continue;
        setClauses.push("\"".concat(col, "\" = ?"));
        params.push(toSqliteValue(value));
    }
    return { sql: setClauses.join(', '), params: params };
}
// ═══════════════════════════════════════════════════════════════════════════════
// Members — 读操作
// ═══════════════════════════════════════════════════════════════════════════════
export function listMembers() {
    var sqlite = tryGetSqlite();
    if (!sqlite)
        return null;
    try {
        var rows = sqlite.prepare('SELECT * FROM members ORDER BY created_at DESC').all();
        return rows.map(rowToCamel);
    }
    catch (err) {
        log.error('[SQLite] members.list error:', err);
        return null;
    }
}
// ═══════════════════════════════════════════════════════════════════════════════
// Members — 写操作
// ═══════════════════════════════════════════════════════════════════════════════
export function createMember(member) {
    var _a;
    var sqlite = tryGetSqlite();
    if (!sqlite)
        return false;
    try {
        var params = toInsertParams(MEM_COLUMNS, MEM_INSERT_COLS, member);
        (_a = sqlite.prepare(MEM_INSERT_SQL)).run.apply(_a, params);
        return true;
    }
    catch (err) {
        log.error('[SQLite] members.create error:', err);
        return false;
    }
}
export function updateMember(member) {
    var _a;
    var sqlite = tryGetSqlite();
    if (!sqlite)
        return false;
    try {
        var _b = toUpdateSet(MEM_COLUMNS, member, ['id']), setSql = _b.sql, setParams = _b.params;
        if (!setSql)
            return true;
        setParams.push(member.id);
        var result = (_a = sqlite.prepare("UPDATE members SET ".concat(setSql, " WHERE id = ?"))).run.apply(_a, setParams);
        return result.changes > 0;
    }
    catch (err) {
        log.error('[SQLite] members.update error:', err);
        return false;
    }
}
export function deleteMember(id) {
    var sqlite = tryGetSqlite();
    if (!sqlite)
        return false;
    try {
        sqlite.prepare('DELETE FROM members WHERE id = ?').run(id);
        return true;
    }
    catch (err) {
        log.error('[SQLite] members.delete error:', err);
        return false;
    }
}
/** 创建薪资历史记录（member create 时的副作用） */
export function createSalaryHistory(entry) {
    var _a;
    var sqlite = tryGetSqlite();
    if (!sqlite)
        return false;
    try {
        sqlite.prepare("\n      INSERT INTO salary_history (member_id, effective_date, base_salary, subsidy, subsidy_note, note, created_at)\n      VALUES (?, ?, ?, ?, ?, ?, ?)\n    ").run(entry.memberId, entry.effectiveDate, toSqliteValue(entry.baseSalary), toSqliteValue((_a = entry.subsidy) !== null && _a !== void 0 ? _a : 0), entry.subsidyNote || '', entry.note || '', entry.createdAt || new Date().toISOString());
        return true;
    }
    catch (err) {
        log.error('[SQLite] salaryHistory.create error:', err);
        return false;
    }
}
// ═══════════════════════════════════════════════════════════════════════════════
// Worker Teams — 读操作
// ═══════════════════════════════════════════════════════════════════════════════
export function listTeams() {
    var sqlite = tryGetSqlite();
    if (!sqlite)
        return null;
    try {
        var rows = sqlite.prepare("\n      SELECT t.*,\n        p.name as project_name,\n        m.name as leader_name\n      FROM worker_teams t\n      LEFT JOIN projects p ON t.project_id = p.id\n      LEFT JOIN members m ON t.leader_id = m.id\n      ORDER BY t.created_at DESC\n    ").all();
        return rows.map(function (row) {
            var camel = rowToCamel(row);
            camel.projectName = row.project_name || '';
            camel.leaderName = row.leader_name || '';
            return camel;
        });
    }
    catch (err) {
        log.error('[SQLite] workerTeams.list error:', err);
        return null;
    }
}
// ═══════════════════════════════════════════════════════════════════════════════
// Worker Teams — 写操作
// ═══════════════════════════════════════════════════════════════════════════════
export function createTeam(team) {
    var _a;
    var sqlite = tryGetSqlite();
    if (!sqlite)
        return false;
    try {
        var params = toInsertParams(TEAM_COLUMNS, TEAM_INSERT_COLS, team);
        (_a = sqlite.prepare(TEAM_INSERT_SQL)).run.apply(_a, params);
        return true;
    }
    catch (err) {
        log.error('[SQLite] workerTeams.create error:', err);
        return false;
    }
}
export function updateTeam(team) {
    var _a;
    var sqlite = tryGetSqlite();
    if (!sqlite)
        return false;
    try {
        var _b = toUpdateSet(TEAM_COLUMNS, team, ['id']), setSql = _b.sql, setParams = _b.params;
        // 始终更新 updated_at
        setParams.push(new Date().toISOString());
        setParams.push(team.id);
        var result = (_a = sqlite.prepare("UPDATE worker_teams SET ".concat(setSql, ", \"updated_at\" = ? WHERE id = ?"))).run.apply(_a, setParams);
        return result.changes > 0;
    }
    catch (err) {
        log.error('[SQLite] workerTeams.update error:', err);
        return false;
    }
}
export function deleteTeam(id) {
    var sqlite = tryGetSqlite();
    if (!sqlite)
        return false;
    try {
        sqlite.prepare('DELETE FROM worker_teams WHERE id = ?').run(id);
        return true;
    }
    catch (err) {
        log.error('[SQLite] workerTeams.delete error:', err);
        return false;
    }
}
// ═══════════════════════════════════════════════════════════════════════════════
// Worker Transfer Records — 操作
// ═══════════════════════════════════════════════════════════════════════════════
export function listTransferRecords(workerId) {
    var _a;
    var sqlite = tryGetSqlite();
    if (!sqlite)
        return null;
    try {
        var sql = 'SELECT * FROM worker_transfer_records';
        var params = [];
        if (workerId) {
            sql += ' WHERE worker_id = ?';
            params.push(workerId);
        }
        sql += ' ORDER BY transfer_date DESC';
        var rows = (_a = sqlite.prepare(sql)).all.apply(_a, params);
        return rows.map(rowToCamel);
    }
    catch (err) {
        log.error('[SQLite] transferRecords.list error:', err);
        return null;
    }
}
export function createTransferRecord(record) {
    var _a;
    var sqlite = tryGetSqlite();
    if (!sqlite)
        return false;
    try {
        var params = toInsertParams(TRANSFER_COLUMNS, TRANSFER_INSERT_COLS, record);
        (_a = sqlite.prepare(TRANSFER_INSERT_SQL)).run.apply(_a, params);
        return true;
    }
    catch (err) {
        log.error('[SQLite] transferRecords.create error:', err);
        return false;
    }
}
// ═══════════════════════════════════════════════════════════════════════════════
// Project Members — 操作
// ═══════════════════════════════════════════════════════════════════════════════
export function listProjectMembers(projectId) {
    var sqlite = tryGetSqlite();
    if (!sqlite)
        return null;
    try {
        var rows = sqlite.prepare("\n      SELECT pm.*, m.name as member_name, m.phone as member_phone, m.role as member_role,\n        m.member_type, m.id_card as member_id_card\n      FROM project_members pm\n      LEFT JOIN members m ON pm.member_id = m.id\n      WHERE pm.project_id = ?\n      ORDER BY pm.joined_at DESC\n    ").all(projectId);
        return rows.map(function (row) {
            var camel = rowToCamel(row);
            // 构建 member 子对象，供前端使用
            camel.member = {
                id: camel.memberId,
                name: row.member_name || '',
                phone: row.member_phone || '',
                role: row.member_role || '',
                memberType: camel.memberType,
                idCard: row.member_id_card || '',
            };
            // 清理冗余字段
            delete camel.member_name;
            delete camel.member_phone;
            delete camel.member_role;
            delete camel.member_id_card;
            return camel;
        });
    }
    catch (err) {
        log.error('[SQLite] projectMembers.list error:', err);
        return null;
    }
}
export function addProjectMember(entry) {
    var _a;
    var sqlite = tryGetSqlite();
    if (!sqlite)
        return false;
    try {
        var params = toInsertParams(PM_COLUMNS, PM_INSERT_COLS, entry);
        (_a = sqlite.prepare(PM_INSERT_SQL)).run.apply(_a, params);
        return true;
    }
    catch (err) {
        log.error('[SQLite] projectMembers.add error:', err);
        return false;
    }
}
export function updateProjectMember(id, changes) {
    var _a;
    var sqlite = tryGetSqlite();
    if (!sqlite)
        return false;
    try {
        var _b = toUpdateSet(PM_COLUMNS, changes, ['id', 'projectId', 'memberId']), setSql = _b.sql, setParams = _b.params;
        if (!setSql)
            return true;
        setParams.push(id);
        var result = (_a = sqlite.prepare("UPDATE project_members SET ".concat(setSql, " WHERE id = ?"))).run.apply(_a, setParams);
        return result.changes > 0;
    }
    catch (err) {
        log.error('[SQLite] projectMembers.update error:', err);
        return false;
    }
}
export function removeProjectMember(id) {
    var sqlite = tryGetSqlite();
    if (!sqlite)
        return false;
    try {
        sqlite.prepare('DELETE FROM project_members WHERE id = ?').run(id);
        return true;
    }
    catch (err) {
        log.error('[SQLite] projectMembers.remove error:', err);
        return false;
    }
}

File: src-tauri/src/commands/members.rs
/**
 * 成员管理命令
 *
 * 对应 Electron 版本的 members.ts
 * 包含成员 CRUD、班组管理、项目成员关联
 */

use crate::db::AppState;
use crate::error::{AppError, AppResult};
use rusqlite::params;
use serde::{Deserialize, Serialize};
use tauri::{command, State};

// ============ 成员类型 ============

#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Member {
    pub id: i64,
    pub name: String,
    pub phone: Option<String>,
    pub email: Option<String>,
    pub member_type: Option<String>, // 'staff' | 'worker'
    pub role: Option<String>,
    pub worker_type: Option<String>,
    pub id_card: Option<String>,
    pub gender: Option<String>,
    pub ethnicity: Option<String>,
    pub birth_date: Option<String>,
    pub id_card_address: Option<String>,
    pub base_salary: Option<f64>,
    pub daily_wage: Option<f64>,
    pub entry_date: Option<String>,
    pub status: Option<String>,
    pub department_id: Option<i64>,
    pub position: Option<String>,
    pub created_at: Option<String>,
    // 关联查询时附带
    pub team_name: Option<String>,
    pub project_name: Option<String>,
    pub department_name: Option<String>,
}

/// 获取所有成员
#[command]
pub fn get_members(state: State<'_, AppState>) -> AppResult<Vec<Member>> {
    let db = state.db.lock().map_err(|e| AppError::Database(e.to_string()))?;

    let mut stmt = db
        .prepare(
            "SELECT m.id, m.name, m.phone, m.email, m.member_type, m.role,
                    m.worker_type, m.id_card, m.gender, m.ethnicity, m.birth_date,
                    m.id_card_address, m.base_salary, m.daily_wage, m.entry_date,
                    m.status, m.department_id, m.position, m.created_at,
                    wt.name as team_name,
                    p.name as project_name,
                    d.name as department_name
             FROM members m
             LEFT JOIN worker_teams wt ON m.team_id = wt.id
             LEFT JOIN projects p ON wt.project_id = p.id
             LEFT JOIN departments d ON m.department_id = d.id
             ORDER BY m.created_at DESC",
        )
        .map_err(|e| AppError::Database(e.to_string()))?;

    let members = stmt
        .query_map([], |row| {
            Ok(Member {
                id: row.get(0)?,
                name: row.get(1)?,
                phone: row.get(2)?,
                email: row.get(3)?,
                member_type: row.get(4)?,
                role: row.get(5)?,
                worker_type: row.get(6)?,
                id_card: row.get(7)?,
                gender: row.get(8)?,
                ethnicity: row.get(9)?,
                birth_date: row.get(10)?,
                id_card_address: row.get(11)?,
                base_salary: row.get(12)?,
                daily_wage: row.get(13)?,
                entry_date: row.get(14)?,
                status: row.get(15)?,
                department_id: row.get(16)?,
                position: row.get(17)?,
                created_at: row.get(18)?,
                team_name: row.get(19)?,
                project_name: row.get(20)?,
                department_name: row.get(21)?,
            })
        })
        .map_err(|e| AppError::Database(e.to_string()))?
        .collect::<Result<Vec<_>, _>>()
        .map_err(|e| AppError::Database(e.to_string()))?;

    Ok(members)
}

/// 创建成员
#[command]
pub fn create_member(state: State<'_, AppState>, member: Member) -> AppResult<i64> {
    let db = state.db.lock().map_err(|e| AppError::Database(e.to_string()))?;

    db.execute(
        "INSERT INTO members (name, phone, email, member_type, role, worker_type,
         id_card, gender, ethnicity, birth_date, id_card_address, base_salary,
         daily_wage, entry_date, status, department_id, position)
         VALUES (?1, ?2, ?3, ?4, ?5, ?6, ?7, ?8, ?9, ?10, ?11, ?12, ?13, ?14, ?15, ?16, ?17)",
        params![
            member.name,
            member.phone,
            member.email,
            member.member_type,
            member.role,
            member.worker_type,
            member.id_card,
            member.gender,
            member.ethnicity,
            member.birth_date,
            member.id_card_address,
            member.base_salary,
            member.daily_wage,
            member.entry_date,
            member.status,
            member.department_id,
            member.position,
        ],
    )
    .map_err(|e| AppError::Database(e.to_string()))?;

    let id = db.last_insert_rowid();

    // 如果有基本工资，自动创建薪资历史记录
    if let Some(base_salary) = member.base_salary {
        if base_salary > 0.0 {
            let effective_date = member
                .entry_date
                .unwrap_or_else(|| chrono::Local::now().format("%Y-%m-%d").to_string());

            db.execute(
                "INSERT INTO salary_history (member_id, effective_date, base_salary, note)
                 VALUES (?1, ?2, ?3, '入职初始薪资')",
                params![id, effective_date, base_salary],
            )
            .map_err(|e| AppError::Database(e.to_string()))?;
        }
    }

    Ok(id)
}

/// 更新成员
#[command]
pub fn update_member(state: State<'_, AppState>, member: Member) -> AppResult<()> {
    let db = state.db.lock().map_err(|e| AppError::Database(e.to_string()))?;

    let affected = db
        .execute(
            "UPDATE members SET name = ?1, phone = ?2, email = ?3, member_type = ?4,
             role = ?5, worker_type = ?6, id_card = ?7, gender = ?8, ethnicity = ?9,
             birth_date = ?10, id_card_address = ?11, base_salary = ?12, daily_wage = ?13,
             entry_date = ?14, status = ?15, department_id = ?16, position = ?17
             WHERE id = ?18",
            params![
                member.name,
                member.phone,
                member.email,
                member.member_type,
                member.role,
                member.worker_type,
                member.id_card,
                member.gender,
                member.ethnicity,
                member.birth_date,
                member.id_card_address,
                member.base_salary,
                member.daily_wage,
                member.entry_date,
                member.status,
                member.department_id,
                member.position,
                member.id,
            ],
        )
        .map_err(|e| AppError::Database(e.to_string()))?;

    if affected == 0 {
        return Err(AppError::NotFound(format!("成员 {} 不存在", member.id)));
    }

    Ok(())
}

/// 删除成员
#[command]
pub fn delete_member(state: State<'_, AppState>, id: i64) -> AppResult<()> {
    let db = state.db.lock().map_err(|e| AppError::Database(e.to_string()))?;

    let affected = db
        .execute("DELETE FROM members WHERE id = ?1", params![id])
        .map_err(|e| AppError::Database(e.to_string()))?;

    if affected == 0 {
        return Err(AppError::NotFound(format!("成员 {} 不存在", id)));
    }

    // 删除相关的薪资历史
    db.execute(
        "DELETE FROM salary_history WHERE member_id = ?1",
        params![id],
    )
    .map_err(|e| AppError::Database(e.to_string()))?;

    Ok(())
}
